testing/ui-tests/src/utils/pallet_metadata_test_runner.rs

```diff
@@ -0,0 +1,101 @@
+// Copyright 2019-2023 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+use codec::{Decode, Encode};
+use frame_metadata::{RuntimeMetadataPrefixed, RuntimeMetadataV14};
+use std::io::Read;
+use subxt_metadata::retain_metadata_pallets;
+
+static TEST_DIR_PREFIX: &str = "subxt_generated_pallets_ui_tests_";
+static METADATA_FILE: &str = "../../artifacts/polkadot_metadata.scale";
+
+pub struct PalletMetadataTestRunner {
+    metadata: RuntimeMetadataV14,
+    index: usize,
+}
+
+impl PalletMetadataTestRunner {
+    pub fn new() -> PalletMetadataTestRunner {
+        let mut file =
+            std::fs::File::open(METADATA_FILE).expect("Cannot open metadata.scale artifact");
+
+        let mut bytes = Vec::new();
+        file.read_to_end(&mut bytes)
+            .expect("Failed to read metadata.scale file");
+
+        let meta: RuntimeMetadataPrefixed =
+            Decode::decode(&mut &*bytes).expect("Cannot decode metadata bytes");
+
+        let metadata = match meta.1 {
+            frame_metadata::RuntimeMetadata::V14(v14) => v14,
+            _ => panic!("Unsupported metadata version. Tests support only v14"),
+        };
+
+        PalletMetadataTestRunner { metadata, index: 0 }
+    }
+
+    pub fn path_to_next_ui_test(&mut self) -> Option<String> {
+        let Some(pallet) = self.metadata.pallets.get(self.index) else {
+            return None
+        };
+        let test_name = &pallet.name;
+
+        // Increment test index to avoid overlaps.
+        let index = self.index;
+        self.index += 1;
+
+        // Build custom metadata containing only this pallet.
+        let mut metadata = self.metadata.clone();
+        retain_metadata_pallets(&mut metadata, |pallet_filter| {
+            pallet_filter.name == pallet.name
+        });
+
+        let mut tmp_dir = std::env::temp_dir();
+        tmp_dir.push(format!("{TEST_DIR_PREFIX}{index}"));
+
+        let tmp_metadata_path = {
+            let mut t = tmp_dir.clone();
+            t.push("metadata.scale");
+            t.to_string_lossy().into_owned()
```

Comment:
probably fine to unwrap here as well because I would regard `std::env::temp_dir()` to generate a path with non-unicode characters as a bug.

the dir wouldn't be found in those cases anyway but yeah it would be catched below `create::dir(&tmp_dir)`